testing/trino-tests/src/test/java/io/trino/tests/TestRejectTaskOnNodeRestart.java

```diff
@@ -0,0 +1,118 @@
+
+import io.trino.execution.QueryManager;
+import io.trino.server.BasicQueryInfo;
+import io.trino.server.testing.TestingTrinoServer;
+import io.trino.testing.DistributedQueryRunner;
+import io.trino.testing.MaterializedResult;
+import io.trino.tests.tpch.TpchQueryRunner;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.parallel.Execution;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.function.Predicate;
+
+import static com.google.common.collect.MoreCollectors.onlyElement;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.trino.execution.QueryState.RUNNING;
+import static io.trino.testing.assertions.Assert.assertEventually;
+import static java.util.UUID.randomUUID;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.parallel.ExecutionMode.SAME_THREAD;
+
+/**
+ * Test that tasks are cleanly rejected when a node restarts (same node ID, different node instance ID).
+ */
+@Execution(SAME_THREAD) // run single threaded to avoid creating multiple query runners at once
+public class TestRejectTaskOnNodeRestart
+{
+    @Test
+    @Timeout(30)
+    public void testNodeRestartBeforeQueryStart()
+            throws Exception
+    {
+        test(true);
+    }
+
+    @Test
+    @Timeout(30)
+    public void testNodeRestartDuringQueryExecution()
+            throws Exception
+    {
+        test(false);
+    }
+
+    private void test(boolean nodeRestartBeforeQueryStart)
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = TpchQueryRunner.builder().build();
+                ExecutorService executor = newCachedThreadPool(daemonThreadsNamed(getClass().getSimpleName() + "-%d"))) {
+            try {
+                TestingTrinoServer worker = queryRunner.getServers().stream()
+                        .filter(Predicate.not(TestingTrinoServer::isCoordinator))
+                        .findFirst()
+                        .orElseThrow();
+
+                // Ensure everything initialized
+                assertThat((long) queryRunner.execute("SELECT count(*) FROM tpch.tiny.lineitem").getOnlyValue())
+                        .isEqualTo(60_175);
+
+                if (nodeRestartBeforeQueryStart) {
+                    queryRunner.restartWorker(worker);
+                }
+
+                String sql = "SELECT count(*) FROM tpch.sf1000000000.lineitem -- " + randomUUID();
+                Future<MaterializedResult> future = executor.submit(() -> queryRunner.execute(sql));
+
+                if (!nodeRestartBeforeQueryStart) {
+                    waitForQueryStart(queryRunner, sql);
+                    queryRunner.restartWorker(worker);
+                }
+
+                assertThatThrownBy(future::get)
```

Comment:
why do you expect error if `nodeRestartBeforeQueryStart==true`?

Reply:
because coordinator's copy of services isn't updated yet, and coordinator tries to schedule a new task on an old instance of a worker. This is exactly the kind of production problem that sparked this PR -- worker was restarted after crash and coordinator was trying to use this new worker before it's initialized (has catalogs) because it was thinking it's the previously announced instance.

in my local testing `testNodeRestartBeforeQueryStart` passed 50 of 50.

Reply:
To your point, there is no strong guarantee, as there is no synchronization / delayed registration for a newly started worker. Moreover, the worker startup is synchronous.

Reply:
Ok so the assumption here is that we are scheduling query quickly enough that coordinator state is stale, right?

Reply:
BTW this indicates a bigger problem:
the worker crash and restart has a more prolonged effect on query execution that it could.
I.e. it affects currently executing queries, but also new queries.
A user can retry a query after it failed due to a worker crash, and the retried query can still fail because of that very same worker crash.

@electrum @dain is this a known limitation?

Reply:
i run the test removing the instance_id related changes (both: checks in TaskResource and discovery propagation).

The results for `testRestartDuringQuery` were better before changes in this PR in terms on how quickly the cluster heals itself.
After worker is started, the next query succeeds. But it doesn't mean a retried query is guaranteed to succeed. It won't succeed, if retry happens before the worker is fully restarted.

The results for `testRestartBeforeQuery` were much worse before changes in this PR.
The worker crash pretty deterministically leads to next query hanging indefinitely. This also means that in `testRestartDuringQuery` case, if the retried query was sent too quickly (before worker fully restarted), it would hang instead of failing.

Thus, I consider this PR an improvement and want to continue on it & get it merged.
But, we may want to iterate further and heal from crashes faster.


cc @losipiuk @electrum

Reply:
> The results for `testRestartBeforeQuery` were much worse before changes in this PR.
> The worker crash pretty deterministically leads to next query hanging indefinitely.

that's was just observation mistake.
my test query was simply never completing (very long), but executing just fine.